src/storage/store/engine.rs

```diff
@@ -1472,10 +1520,33 @@ impl ShardEngine {
         let trie_ctx = &merkle_trie::Context::with_callback(count_callback);
 
         let mut applied_cached_txn = false;
-        if let Some(cached_txn) = self.pending_txn.clone() {
+        if let Some(mut cached_txn) = self.pending_txn.clone() {
             if &cached_txn.shard_root == shard_root {
                 applied_cached_txn = true;
-                self.commit_and_emit_events(shard_chunk, cached_txn.events, cached_txn.txn);
+                // BLOCK_CONFIRMED Integration - Cached Transaction Path:
+                // When we have a valid cached transaction (from validation), we add the
+                // BLOCK_CONFIRMED event to the existing events before committing.
+                // This ensures every block gets a BLOCK_CONFIRMED event regardless of
+                // whether it goes through the full replay or uses cached results.
+                match self.add_block_confirmed_event(
+                    shard_chunk,
+                    cached_txn.events,
+                    &mut cached_txn.txn,
+                ) {
+                    Ok(events_with_block_confirmed) => {
+                        self.commit_and_emit_events(
```

Comment:
Can you move `add_block_confirmed_event` to commit_and_emit_events given it's called before that in both places?